src/librbd/ImageWatcher.cc

```diff
@@ -489,6 +489,46 @@ void ImageWatcher<I>::notify_lock_owner(const Payload& payload,
   notify_lock_owner->send();
 }
 
+template <typename I>
+bool ImageWatcher<I>::is_new_request(const AsyncRequestId &id) const {
+  ceph_assert(ceph_mutex_is_locked(m_async_request_lock));
+
+  return m_async_pending.count(id) == 0 && m_async_complete.count(id) == 0;
+}
+
+template <typename I>
+bool ImageWatcher<I>::mark_async_request_complete(const AsyncRequestId &id,
+                                                  int r) {
+  ceph_assert(ceph_mutex_is_locked(m_async_request_lock));
+
+  bool found = m_async_pending.erase(id);
+
+  auto now = ceph_clock_now();
+
+  auto it = m_async_complete_expiration.begin();
+  while (it != m_async_complete_expiration.end() && it->first < now) {
+    m_async_complete.erase(it->second);
+    it = m_async_complete_expiration.erase(it);
+  }
+
+  if (!m_async_complete.insert({id, r}).second) {
+    for (it = m_async_complete_expiration.begin();
+         it != m_async_complete_expiration.end(); it++) {
+      if (it->second != id) {
+        continue;
+      } else {
+        m_async_complete_expiration.erase(it);
+        break;
+      }
```

Comment:
```suggestion
      if (!(it->second != id)) {
        m_async_complete_expiration.erase(it);
        break;
      }
```

Reply:
Actually it was my version but it looked unreadable and might make wonder an occasional reader.  But ok. I will use it.